traitlets/traitlets.py

```diff
@@ -1129,37 +1182,38 @@ def hold(change):
                         self.notify_change(change)
 
     def _notify_trait(self, name, old_value, new_value):
-        self.notify_change(Bunch(
+        self.notify_change(Change(
             name=name,
             old=old_value,
             new=new_value,
             owner=self,
-            type='change',
         ))
 
     def notify_change(self, change):
-        if not isinstance(change, Bunch):
+        if not isinstance(change, EventReport):
             # cast to bunch if given a dict
-            change = Bunch(change)
-        name, type = change.name, change.type
+            change = EventReport(change)
 
         callables = []
-        callables.extend(self._trait_notifiers.get(name, {}).get(type, []))
-        callables.extend(self._trait_notifiers.get(name, {}).get(All, []))
-        callables.extend(self._trait_notifiers.get(All, {}).get(type, []))
-        callables.extend(self._trait_notifiers.get(All, {}).get(All, []))
+        if 'name' in change:
+            # name is not gaurenteed to be in the event report
```

Comment:
Maybe add to the docs that `'name'` is required instead of handling the never-true case here, since there are currently no cases where it can be absent, and there probably shouldn't ever be any.